rat/codegen_backends.py

```diff
@@ -1,357 +1,269 @@
+
+    def walk_Binary(self, node: ast.Binary):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        return eval(f"left {node.op} right")
+
+    def walk_IfElse(self, node: ast.IfElse):
+        predicate = self.arguments[node.predicate]
+        return jax.lax.cond(predicate, lambda: self.walk(node.left), lambda: self.walk(node.right))
+
+    def walk_FunctionCall(self, node: ast.FunctionCall):
+        arglist = []
+
+        if self.left_side_of_sampling:
+            arglist += [self.walk(self.left_side_of_sampling)]
+
+        if node.arglist:
+            arglist += [self.walk(arg) for arg in node.arglist]
+
+        return eval(node.name)(*arglist)
+
+    def walk_Variable(self, node: ast.Variable):
+        if node in self.arguments:
+            if node.name in self.parameters:
+                trace = self.arguments[node]
+                return self.parameters[node.name][trace]
+            else:
+                return self.arguments[node]
+        else:
+            return self.parameters[node.name]
+
+    def walk_Literal(self, node: ast.Literal):
+        return node.value
+
+
+class TraceExecutor(NodeWalker):
+    """
+    As long as shunt is true, values of expressions aren't returned but
+    instead shunted off into tuples
+    """
+
+    tracers: Dict[str, Tracer]
+    leaves: Dict[str, Any]
+    shunt: ContextStack[bool]
+    trace_by_reference: Set[ast.Variable]
+    first_level_values: Dict[ast.Variable, Any]
+
+    def __init__(self, tracers: Dict[str, Tracer], leaves: Dict[str, Any], trace_by_reference: Set[ast.Variable] = None):
+        self.tracers = tracers
+        self.leaves = leaves
+        self.shunt = ContextStack(True)
+        self.trace_by_reference = trace_by_reference if trace_by_reference else set()
+        self.first_level_values = {}
+
+    def walk_Statement(self, node: ast.Statement):
+        self.walk(node.left)
+        self.walk(node.right)
+
+    def walk_Logical(self, node: ast.Logical):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        if not self.shunt.peek():
+            return eval(f"left {node.op} right")
+
+    def walk_Binary(self, node: ast.Binary):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        if not self.shunt.peek():
+            return eval(f"left {node.op} right")
+
+    def walk_IfElse(self, node: ast.IfElse):
+        with self.shunt.push(False):
+            predicate = self.walk(node.predicate)
+
+        if self.shunt.peek():
+            self.first_level_values[node] = predicate
+
+        return self.walk(node.left) if predicate else self.walk(node.right)
+
+    def walk_FunctionCall(self, node: ast.FunctionCall):
+        if node.arglist:
+            arglist = tuple(self.walk(arg) for arg in node.arglist)
+        else:
+            arglist = ()
+
+        if not self.shunt.peek():
+            return eval(node.name)(*arglist)
```

Comment:
The math functions in rat are defined simply as `log`, `sin`, but looks like this would fail because instead it should be calling `jax.numpy.log` and so on.

Reply:
Good point. Yah boy will just sneak in a from jax.numpy import * for now.

Reply:
It's totally possible we do something clean here that isn't just importing namespaces and hoping they don't clobber each other. I feeling is we wait until we allow people to pass in custom user defined python functions and bind them to names in Rat to worry about this tho.